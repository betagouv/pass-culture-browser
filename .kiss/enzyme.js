import React from 'react'
import { shallow } from 'enzyme'

import MyComponent from '../MyComponent'

describe('src | components | MyComponent', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      const props = { errors: {} }
      const wrapper = shallow(<MyComponent {...props} />)
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
