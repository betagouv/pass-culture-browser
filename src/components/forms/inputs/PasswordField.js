/* eslint
  react/jsx-one-expression-per-line: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'

import FormError from '../FormError'
import InputHelp from '../InputHelp'
import InputLabel from '../InputLabel'
import { validatePasswordField } from '../validators'

class PasswordField extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = { hidden: true }
  }

  onToggleVivisbility = () => {
    this.setState(prev => ({ hidden: !prev.hidden }))
  }

  render() {
    const {
      autoComplete,
      className,
      disabled,
      help,
      id,
      label,
      name,
      placeholder,
      required,
      sublabel,
    } = this.props
    const { hidden } = this.state
    const status = hidden ? '' : '-close'
    const inputType = hidden ? 'password' : 'text'
    const validateFunc =
      required && typeof required === 'function'
        ? required
        : (required && validatePasswordField) || undefined
    return (
      <Field
        name={name}
        // fallback to form validator
        // si le champs n'est pas marquÃ© comme requis
        render={({ input, meta }) => (
          <p className={`${className}`}>
            <label
              className="pc-final-form-password"
              htmlFor={id || name}
            >
              {label && (
                <InputLabel
                  label={label}
                  required={required}
                  sublabel={sublabel}
                />
              )}
              {help && <InputHelp label={help} />}
              <span className="pc-final-form-inner">
                <input
                  {...input}
                  autoComplete={autoComplete ? 'on' : 'off'}
                  className="pc-final-form-input"
                  disabled={disabled}
                  id={id || name} // cast to boolean
                  placeholder={placeholder}
                  required={!!required}
                  type={inputType}
                />

                <button
                  className="no-border no-outline no-background mx12 flex-0 is-primary-text"
                  onClick={this.onToggleVivisbility}
                  type="button"
                >
                  <span
                    aria-hidden
                    className={`icon-legacy-eye${status} fs22`}
                    title="Afficher/Masquer le mot de passe"
                  />
                </button>
              </span>
              <FormError
                id={`${id || name}-error`}
                meta={meta}
              />
            </label>
          </p>
        )}
        validate={validateFunc || undefined}
      />
    )
  }
}

PasswordField.defaultProps = {
  autoComplete: false,
  className: '',
  disabled: false,
  help: null,
  id: null,
  label: 'Saisissez Votre mot de passe',
  placeholder: '',
  required: false,
  sublabel: null,
}

PasswordField.propTypes = {
  autoComplete: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  help: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  sublabel: PropTypes.string,
}

export default PasswordField
