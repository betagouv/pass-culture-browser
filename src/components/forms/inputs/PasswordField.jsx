import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'

import FormError from '../FormError'
import InputLabel from '../InputLabel'
import { validatePasswordField } from '../validators'
import hasErrorMessage from '../utils/hasErrorMessage'

class PasswordField extends PureComponent {
  constructor(props) {
    super(props)
    this.state = { hidden: true }
  }

  handleOnToggleVisibility = () => {
    this.setState(prev => ({ hidden: !prev.hidden }))
  }

  renderField = ({ input, meta }) => {
    const { autoComplete, disabled, id, label, name, placeholder, required, sublabel } = this.props
    const { hidden } = this.state
    const status = hidden ? '' : '-close'
    const inputType = hidden ? 'password' : 'text'

    return (
      <label className="label-password-inner">
        <InputLabel
          label={label}
          required={required}
          sublabel={sublabel}
        />
        <div className={`input-inner${hasErrorMessage(meta)}`}>
          <input
            {...input}
            autoComplete={autoComplete ? 'on' : 'off'}
            className="form-input"
            disabled={disabled}
            id={id || name}
            placeholder={placeholder}
            required={!!required}
            type={inputType}
          />
          <button
            className="no-background mx12 flex-0 is-primary-text"
            onClick={this.handleOnToggleVisibility}
            type="button"
          >
            <span
              aria-hidden
              className={`icon-legacy-eye${status}`}
              title="Afficher/Masquer le mot de passe"
            />
          </button>
        </div>
        <FormError meta={meta} />
      </label>
    )
  }

  render() {
    const { name, required } = this.props
    const validateFunc =
      required && typeof required === 'function'
        ? required
        : (required && validatePasswordField) || undefined

    return (
      <Field
        name={name}
        // fallback to form validator
        // si le champs n'est pas marquÃ© comme requis
        render={this.renderField}
        validate={validateFunc || undefined}
      />
    )
  }
}

PasswordField.defaultProps = {
  autoComplete: false,
  disabled: false,
  id: null,
  placeholder: '',
  required: false,
  sublabel: null,
}

PasswordField.propTypes = {
  autoComplete: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  sublabel: PropTypes.string,
}

export default PasswordField
