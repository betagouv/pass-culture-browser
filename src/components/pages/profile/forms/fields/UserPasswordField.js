import PropTypes from 'prop-types'
import React from 'react'

import { FormError } from '../../../../forms'
import { isEmpty } from '../../../../../utils/strings'
import { PasswordField } from '../../../../forms/inputs'
import { validateMatchingFields } from '../../../../forms/validators'
import withProfileForm from '../withProfileForm'

const ERROR_OLD_PASSWORD = "L'ancien mot de passe est manquant."

export class UserPasswordField extends React.PureComponent {
  buildOldPasswordLabel = () => value =>
    value && !isEmpty(value) ? undefined : ERROR_OLD_PASSWORD

  validateNewPassword = () => (newPasswordConfirm, formvalues) => {
    const { newPassword } = formvalues
    return validateMatchingFields(newPasswordConfirm, newPassword)
  }

  render() {
    const { formErrors, isLoading } = this.props
    return (
      <div className="pc-scroll-container">
        <div className="py30 px12 flex-1">
          <PasswordField
            disabled={isLoading}
            label="Saisissez votre mot de passe actuel"
            name="oldPassword"
            required={this.buildOldPasswordLabel()}
          />
          <PasswordField
            className="mt36"
            disabled={isLoading}
            label="Saisissez votre nouveau mot de passe"
            name="newPassword"
            required
          />
          <PasswordField
            className="mt36"
            disabled={isLoading}
            label="Confirmez votre nouveau mot de passe"
            name="newPasswordConfirm"
            required={this.validateNewPassword()}
          />
          {formErrors && <FormError customMessage={formErrors} />}
        </div>
      </div>
    )
  }
}

UserPasswordField.defaultProps = {
  formErrors: false,
}

UserPasswordField.propTypes = {
  formErrors: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.bool,
    PropTypes.string,
  ]),
  isLoading: PropTypes.bool.isRequired,
}

export default withProfileForm(
  UserPasswordField,
  {
    routeMethod: 'POST',
    routePath: 'users/current/change-password',
  },
  null
)
