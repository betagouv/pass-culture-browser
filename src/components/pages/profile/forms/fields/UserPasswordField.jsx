import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

import FormError from '../../../../forms/FormError'
import isEmpty from '../../../../../utils/strings/isEmpty'
import PasswordField from '../../../../forms/inputs/PasswordField'
import validateMatchingFields from '../../../../forms/validators/validateMatchingFields'
import withProfileForm from '../withProfileForm'

const ERROR_OLD_PASSWORD = 'Lâ€™ancien mot de passe est manquant.'

export class UserPasswordField extends React.PureComponent {
  buildOldPasswordLabel = () => value => (value && !isEmpty(value) ? undefined : ERROR_OLD_PASSWORD)

  validateNewPassword = () => (newPasswordConfirm, formvalues) => {
    const { newPassword } = formvalues

    return validateMatchingFields(newPasswordConfirm, newPassword)
  }

  render() {
    const { formErrors, isLoading } = this.props

    return (
      <Fragment>
        <PasswordField
          disabled={isLoading}
          label="Saisissez votre mot de passe actuel"
          name="oldPassword"
          required={this.buildOldPasswordLabel()}
        />
        <PasswordField
          disabled={isLoading}
          label="Saisissez votre nouveau mot de passe"
          name="newPassword"
          required
        />
        <PasswordField
          disabled={isLoading}
          label="Confirmez votre nouveau mot de passe"
          name="newPasswordConfirm"
          required={this.validateNewPassword()}
        />
        {formErrors && <FormError customMessage={formErrors} />}
      </Fragment>
    )
  }
}

UserPasswordField.defaultProps = {
  formErrors: false,
}

UserPasswordField.propTypes = {
  formErrors: PropTypes.oneOfType([PropTypes.array, PropTypes.bool, PropTypes.string]),
  isLoading: PropTypes.bool.isRequired,
}

export default withProfileForm(
  UserPasswordField,
  {
    routeMethod: 'POST',
    routePath: 'users/current/change-password',
  },
  null
)
