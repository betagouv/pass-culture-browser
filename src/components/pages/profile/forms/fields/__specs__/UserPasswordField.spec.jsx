import React from 'react'
import { shallow } from 'enzyme'
import { UserPasswordField } from '../UserPasswordField'

import FormError from '../../../../../forms/FormError'
import validateMatchingFields from '../../../../../forms/validators/validateMatchingFields'
import PasswordField from '../../../../../forms/inputs/PasswordField'

jest.mock('../../../../../forms/validators/validateMatchingFields')

describe('src | components | pages | profile | forms | fields | UserPasswordField', () => {
  let props

  beforeEach(() => {
    props = {
      isLoading: false,
    }
  })

  it('should match the snapshot', () => {
    // when
    const wrapper = shallow(<UserPasswordField {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  it('should render three UserPasswordField components', () => {
    // when
    const wrapper = shallow(<UserPasswordField {...props} />)

    // then
    const passwordFields = wrapper.find(PasswordField)
    expect(passwordFields).toHaveLength(3)
    expect(passwordFields.at(0).prop('required')).toStrictEqual(expect.any(Function))
    expect(passwordFields.at(0).prop('name')).toBe('oldPassword')
    expect(passwordFields.at(0).prop('disabled')).toBe(false)
    expect(passwordFields.at(0).prop('label')).toBe('Saisissez votre mot de passe actuel')
    expect(passwordFields.at(1).prop('required')).toBe(true)
    expect(passwordFields.at(1).prop('name')).toBe('newPassword')
    expect(passwordFields.at(1).prop('disabled')).toBe(false)
    expect(passwordFields.at(1).prop('label')).toBe('Saisissez votre nouveau mot de passe')
    expect(passwordFields.at(2).prop('required')).toStrictEqual(expect.any(Function))
    expect(passwordFields.at(2).prop('name')).toBe('newPasswordConfirm')
    expect(passwordFields.at(2).prop('disabled')).toBe(false)
    expect(passwordFields.at(2).prop('label')).toBe('Confirmez votre nouveau mot de passe')
  })

  it('should render a FormError component when errors', () => {
    // given
    props.formErrors = 'error1'

    // when
    const wrapper = shallow(<UserPasswordField {...props} />)

    // then
    const formError = wrapper.find(FormError)
    expect(formError).toHaveLength(1)
    expect(formError.prop('customMessage')).toBe('error1')
  })

  describe('functions', () => {
    describe('buildOldPasswordLabel', () => {
      it('should return undefined when old password is provided', () => {
        // given
        const wrapper = shallow(<UserPasswordField {...props} />)

        // when
        const result = wrapper.instance().buildOldPasswordLabel()('fake label')

        // then
        expect(result).toBeUndefined()
      })

      it('should return missing old password message when old password is not provided', () => {
        // given
        const wrapper = shallow(<UserPasswordField {...props} />)

        // when
        const result = wrapper.instance().buildOldPasswordLabel()()

        // then
        expect(result).toBe('Lâ€™ancien mot de passe est manquant.')
      })
    })

    describe('validateNewPassword', () => {
      it('should validate new password when provided', () => {
        // given
        const newPasswordConfirm = 'hello'
        const formValues = {
          newPassword: 'hello',
        }
        const wrapper = shallow(<UserPasswordField {...props} />)

        // when
        wrapper.instance().validateNewPassword()(newPasswordConfirm, formValues)

        // then
        expect(validateMatchingFields).toHaveBeenCalledWith('hello', 'hello')
      })
    })
  })
})
