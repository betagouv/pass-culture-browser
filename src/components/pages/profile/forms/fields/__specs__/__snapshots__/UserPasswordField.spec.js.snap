// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | profile | forms | fields | UserPasswordField should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserPasswordField
    formErrors={false}
    isLoading={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="py30 px12 flex-1"
      >
        <PasswordField
          autoComplete={false}
          className=""
          disabled={false}
          help={null}
          id={null}
          label="Saisissez votre mot de passe actuel"
          name="oldPassword"
          placeholder=""
          required={[Function]}
          sublabel={null}
        />
        <PasswordField
          autoComplete={false}
          className="mt36"
          disabled={false}
          help={null}
          id={null}
          label="Saisissez votre nouveau mot de passe"
          name="newPassword"
          placeholder=""
          required={true}
          sublabel={null}
        />
        <PasswordField
          autoComplete={false}
          className="mt36"
          disabled={false}
          help={null}
          id={null}
          label="Confirmez votre nouveau mot de passe"
          name="newPasswordConfirm"
          placeholder=""
          required={[Function]}
          sublabel={null}
        />
      </div>,
      "className": "pc-scroll-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PasswordField
            autoComplete={false}
            className=""
            disabled={false}
            help={null}
            id={null}
            label="Saisissez votre mot de passe actuel"
            name="oldPassword"
            placeholder=""
            required={[Function]}
            sublabel={null}
          />,
          <PasswordField
            autoComplete={false}
            className="mt36"
            disabled={false}
            help={null}
            id={null}
            label="Saisissez votre nouveau mot de passe"
            name="newPassword"
            placeholder=""
            required={true}
            sublabel={null}
          />,
          <PasswordField
            autoComplete={false}
            className="mt36"
            disabled={false}
            help={null}
            id={null}
            label="Confirmez votre nouveau mot de passe"
            name="newPasswordConfirm"
            placeholder=""
            required={[Function]}
            sublabel={null}
          />,
          false,
        ],
        "className": "py30 px12 flex-1",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": false,
            "className": "",
            "disabled": false,
            "help": null,
            "id": null,
            "label": "Saisissez votre mot de passe actuel",
            "name": "oldPassword",
            "placeholder": "",
            "required": [Function],
            "sublabel": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": false,
            "className": "mt36",
            "disabled": false,
            "help": null,
            "id": null,
            "label": "Saisissez votre nouveau mot de passe",
            "name": "newPassword",
            "placeholder": "",
            "required": true,
            "sublabel": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": false,
            "className": "mt36",
            "disabled": false,
            "help": null,
            "id": null,
            "label": "Confirmez votre nouveau mot de passe",
            "name": "newPasswordConfirm",
            "placeholder": "",
            "required": [Function],
            "sublabel": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="py30 px12 flex-1"
        >
          <PasswordField
            autoComplete={false}
            className=""
            disabled={false}
            help={null}
            id={null}
            label="Saisissez votre mot de passe actuel"
            name="oldPassword"
            placeholder=""
            required={[Function]}
            sublabel={null}
          />
          <PasswordField
            autoComplete={false}
            className="mt36"
            disabled={false}
            help={null}
            id={null}
            label="Saisissez votre nouveau mot de passe"
            name="newPassword"
            placeholder=""
            required={true}
            sublabel={null}
          />
          <PasswordField
            autoComplete={false}
            className="mt36"
            disabled={false}
            help={null}
            id={null}
            label="Confirmez votre nouveau mot de passe"
            name="newPasswordConfirm"
            placeholder=""
            required={[Function]}
            sublabel={null}
          />
        </div>,
        "className": "pc-scroll-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <PasswordField
              autoComplete={false}
              className=""
              disabled={false}
              help={null}
              id={null}
              label="Saisissez votre mot de passe actuel"
              name="oldPassword"
              placeholder=""
              required={[Function]}
              sublabel={null}
            />,
            <PasswordField
              autoComplete={false}
              className="mt36"
              disabled={false}
              help={null}
              id={null}
              label="Saisissez votre nouveau mot de passe"
              name="newPassword"
              placeholder=""
              required={true}
              sublabel={null}
            />,
            <PasswordField
              autoComplete={false}
              className="mt36"
              disabled={false}
              help={null}
              id={null}
              label="Confirmez votre nouveau mot de passe"
              name="newPasswordConfirm"
              placeholder=""
              required={[Function]}
              sublabel={null}
            />,
            false,
          ],
          "className": "py30 px12 flex-1",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoComplete": false,
              "className": "",
              "disabled": false,
              "help": null,
              "id": null,
              "label": "Saisissez votre mot de passe actuel",
              "name": "oldPassword",
              "placeholder": "",
              "required": [Function],
              "sublabel": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoComplete": false,
              "className": "mt36",
              "disabled": false,
              "help": null,
              "id": null,
              "label": "Saisissez votre nouveau mot de passe",
              "name": "newPassword",
              "placeholder": "",
              "required": true,
              "sublabel": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoComplete": false,
              "className": "mt36",
              "disabled": false,
              "help": null,
              "id": null,
              "label": "Confirmez votre nouveau mot de passe",
              "name": "newPasswordConfirm",
              "placeholder": "",
              "required": [Function],
              "sublabel": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
